go/oasis-test-runner/scenario/e2e/runtime/runtime_upgrade.go

```diff
@@ -0,0 +1,287 @@
+	if err := cli.Keymanager.SignPolicy("1", kmPolicyPath, kmPolicySig1Path); err != nil {
+		return err
+	}
+	if err := cli.Keymanager.SignPolicy("2", kmPolicyPath, kmPolicySig2Path); err != nil {
+		return err
+	}
+	if err := cli.Keymanager.SignPolicy("3", kmPolicyPath, kmPolicySig3Path); err != nil {
+		return err
+	}
+
+	sc.Logger.Info("updating KM policy")
+	if err := cli.Keymanager.GenUpdate(sc.nonce, kmPolicyPath, []string{kmPolicySig1Path, kmPolicySig2Path, kmPolicySig3Path}, kmUpdateTxPath); err != nil {
+		return err
+	}
+	if err := cli.Consensus.SubmitTx(kmUpdateTxPath); err != nil {
+		return fmt.Errorf("failed to update KM policy: %w", err)
+	}
+	sc.nonce++
+
+	return nil
+}
+
+func (sc *runtimeUpgradeImpl) Run(childEnv *env.Env) error {
+	ctx := context.Background()
+	cli := cli.New(childEnv, sc.Net, sc.Logger)
+
+	computeWorkers := len(sc.Net.ComputeWorkers())
+
+	clientErrCh, cmd, err := sc.runtimeImpl.start(childEnv)
+	if err != nil {
+		return err
+	}
+	sc.Logger.Info("waiting for client to exit")
+	// Wait for the client to exit.
+	select {
+	case err = <-sc.runtimeImpl.Net.Errors():
+		_ = cmd.Process.Kill()
+	case err = <-clientErrCh:
+	}
+	if err != nil {
+		return err
+	}
+
+	// Generate and update a policy that will allow the new runtime to run.
+	if err = sc.applyUpgradePolicy(childEnv); err != nil {
+		return fmt.Errorf("updating policies: %w", err)
+	}
+
+	// Kill old compute workers.
+	sc.Logger.Info("stopping old runtimes")
+	for i := 0; i < computeWorkers/2; i++ {
+		worker := sc.Net.ComputeWorkers()[i]
+		if err = worker.Stop(); err != nil {
+			return fmt.Errorf("stopping old compute worker: %w", err)
+		}
+	}
+
+	// Start the new compute workers.
+	sc.Logger.Info("starting new runtimes")
+	for i := computeWorkers / 2; i < computeWorkers; i++ {
+		newWorker := sc.Net.ComputeWorkers()[i]
+		if err = newWorker.Start(); err != nil {
+			return fmt.Errorf("starting new compute worker: %w", err)
+		}
+	}
+
+	// Update runtime to include the new enclave identity.
+	sc.Logger.Info("updating runtime descriptor")
+	newRt := sc.Net.Runtimes()[len(sc.Net.Runtimes())-1]
+	newRtDesc := newRt.ToRuntimeDescriptor()
+	newTxPath := filepath.Join(childEnv.Dir(), "register_update_compute_runtime.json")
+	if err = cli.Registry.GenerateRegisterRuntimeTx(sc.nonce, newRtDesc, newTxPath, ""); err != nil {
+		return fmt.Errorf("failed to generate register compute runtime tx: %w", err)
+	}
+	sc.nonce++
+	if err = cli.Consensus.SubmitTx(newTxPath); err != nil {
+		return fmt.Errorf("failed to update compute runtime: %w", err)
+	}
+
+	// Wait for the new nodes to register.
+	sc.Logger.Info("waiting for new compute nodes to register",
+		"num_nodes", sc.Net.NumRegisterNodes(),
+	)
+
+	if err = sc.Net.Controller().WaitNodesRegistered(ctx, sc.Net.NumRegisterNodes()); err != nil {
+		return fmt.Errorf("failed to wait for nodes: %w", err)
+	}
+
+	sc.Logger.Info("wait for few epochs")
+	var waitEpoch epochtime.EpochTime
```

Comment:
I think instead of waiting for some epochs, this can now be replaced by waiting for the Compute workers to be ready (see the `WaitReady` method added in: https://github.com/oasisprotocol/oasis-core/pull/2957). Similarly in the `keymanager-runtime-upgrade` upgrade (which should wait for the newly spawned keymanager workers to be ready) - i think it will still need to wait for 1 more epoch transition after the nodes are ready.